source/common/router/router.h

```diff
@@ -161,8 +162,16 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     }
     return {};
   }
-  const Router::MetadataMatchCriteria* metadataMatchCriteria() const override {
+  const Router::MetadataMatchCriteria* metadataMatchCriteria() override {
     if (route_entry_) {
+      // The request's metadata, if present, takes precedence over the route's.
+      const auto& request_metadata = callbacks_->requestInfo().dynamicMetadata().filter_metadata();
+      const auto filter_it = request_metadata.find(Envoy::Config::MetadataFilters::get().ENVOY_LB);
+      if (filter_it != request_metadata.end()) {
+        metadata_match_ =
+            route_entry_->metadataMatchCriteria()->mergeMatchCriteria(filter_it->second);
```

Comment:
I think this needs to handle the case where `route_entry_->metadataMatchCriteria()` is nullptr and add a test case for that path.